File: 98-validate-binary-search-tree/validate_binary_search_tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #include<climits>
class Solution {
public:
    // 改成int 指针会快挺多的，数据类型的优化不可忽视
    long  last =LONG_MIN; 
    bool isValidBST(TreeNode* root) {
        // 递归中序遍历，结果应是严格升序，即为bst
        // 将上一个值存在last成员变量中即可判断是否为bst
        if(!root){
            return true;
        }
        if (root->left && !isValidBST(root->left)){
            return false;
        }
        if ( root->val <= last){
            return false;
        }
        last=root->val;
        if (root->right && !isValidBST(root->right)){
            return false;
        }
        return true;
    }
};

File: 98-validate-binary-search-tree/validate_binary_search_tree.go
package leetcode98

import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	last := math.MinInt64
	stack := make([]*TreeNode, 0)
	p := root
	for p != nil || len(stack) != 0 {
		if p != nil {
			stack = append(stack, p)
			p = p.Left
		}
		if p == nil && len(stack) != 0 {
			p = stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if last >= p.Val {
				return false
			}
			last = p.Val
			p = p.Right
		}
	}
	return true

}
